openedx/openedx-platform
# Lets ENABLE_ENTERPRISE_INTEGRATION be configurable

Allows an edx-platform instance to disable Enterprise integration via configuration.

**JIRA tickets**: [OSPR-1945](https://openedx.atlassian.net/browse/OSPR-1945)

**Dependencies**: 

Related to https://github.com/edx/configuration/pull/4128

**Sandbox URL**:

* LMS: https://pr16155.sandbox.opencraft.hosting/
* Studio: https://studio-pr16155.sandbox.opencraft.hosting/

Running 425239a.

**Merge deadline**: ASAP

**Testing instructions**:

1. Run the LMS shell, and ensure that the default enterprise integration setting is off by default:
   ```bash
   >>> from openedx.features.enterprise_support.api import enterprise_enabled
   >>> enterprise_enabled()
   False
   ```
1. Update `/edx/app/edxapp/lms.env.json` to add to `"FEATURES"`: 
   ```json
   "ENABLE_ENTERPRISE_INTEGRATION": true,
   ```
1. Restart the LMS shell, and verify that `enterprise_enabled()` now returns `True`.

**Author Notes & Concerns**

We're currently experiencing an error during learner registration via course enrolment, for Ginkgo instances with third-party authentication configured, that sit behind a load balancer and cannot resolve their public Enterprise API URL on https.  The user sees a 500 error.
<details><summary>See traceback</summary>
```
Oct  2 15:50:48 ondemand [service_variant=lms][django.request][env:sandbox] ERROR [ondemand  2723] [base.py:256] - Internal Server Error: /course_modes/choose/course-v1:Cloudera+CN+0/
--
Traceback (most recent call last):
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/newrelic/hooks/framework_django.py", line 527, in wrapper
return wrapped(*args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/generic/base.py", line 71, in view
return self.dispatch(request, *args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 34, in _wrapper
return bound_func(*args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 30, in bound_func
return func.__get__(self, type(self))(*args2, **kwargs2)
File "/edx/app/edxapp/edx-platform/common/djangoapps/course_modes/views.py", line 56, in dispatch
return super(ChooseModeView, self).dispatch(*args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/newrelic/hooks/framework_django.py", line 914, in wrapper
return wrapped(*args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/generic/base.py", line 89, in dispatch
return handler(request, *args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 34, in _wrapper
return bound_func(*args, **kwargs)
File "/edx/app/edxapp/edx-platform/common/djangoapps/third_party_auth/decorators.py", line 87, in inner
return func(request, *args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 30, in bound_func
return func.__get__(self, type(self))(*args2, **kwargs2)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 34, in _wrapper
return bound_func(*args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/contrib/auth/decorators.py", line 22, in _wrapped_view
return view_func(request, *args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 30, in bound_func
return func.__get__(self, type(self))(*args2, **kwargs2)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 34, in _wrapper
return bound_func(*args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 145, in inner
return func(*args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 30, in bound_func
return func.__get__(self, type(self))(*args2, **kwargs2)
File "/edx/app/edxapp/edx-platform/common/djangoapps/course_modes/views.py", line 120, in get
return redirect(self._get_redirect_url_for_audit_enrollment(request, course_id))
File "/edx/app/edxapp/edx-platform/common/djangoapps/course_modes/views.py", line 298, in _get_redirect_url_for_audit_enrollment
enterprise_learner_data = enterprise_api.get_enterprise_learner_data(site=request.site, user=request.user)
File "/edx/app/edxapp/edx-platform/openedx/features/enterprise_support/api.py", line 391, in get_enterprise_learner_data
enterprise_learner_data = EnterpriseApiClient().fetch_enterprise_learner_data(site=site, user=user)
File "/edx/app/edxapp/edx-platform/openedx/features/enterprise_support/api.py", line 205, in fetch_enterprise_learner_data
response = endpoint().get(**querystring)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/slumber/__init__.py", line 155, in get
resp = self._request("GET", params=kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/slumber/__init__.py", line 97, in _request
resp = self._store["session"].request(method, url, data=data, params=params, files=files, headers=headers)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/requests/sessions.py", line 468, in request
resp = self.send(prep, **send_kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/newrelic/api/external_trace.py", line 121, in dynamic_wrapper
return wrapped(*args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/requests/sessions.py", line 576, in send
r = adapter.send(request, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/requests/adapters.py", line 437, in send
raise ConnectionError(e, request=request)
```
</details>

This change allows us to disable Enterprise integration and avoid this issue, but could be addressed in several other ways.

**Reviewers**
- [ ] edX reviewer[s] TBD CC @gsong 

**Settings**
```yaml
EDXAPP_FEATURES:
  ENABLE_ENTERPRISE_INTEGRATION: false
```
